### ISSUE TITLE
Drpout layer:  noise.py: RandomStreams: theano.sandbox.rng_mrg.mrg_uniform doesn't execute on GPU

### ISSUE THREAD

**Issue by jc-Rosichini** [Lasagne/Lasagne#829]
Hi, due to recent changes in Theano and libgpuarray git, it appears that for a net using extensively dropout layers the performances have collapsed by a factor of almost 10, which is a real pain.

Digging into the issue it seems that .rng_mrg.mrg_uniform doesn't execute on GPU, the RandomStreams(get_rng().randint(1, 2147462579),**use_cuda = True**) is no longer possible.

Question: is there a way to fix by using a CuRand implementation or other?
Thanks for some hints.

**Comment by f0k** [comment]
Where do you see a `RandomStreams` construction with `use_cuda=True`? Lasagne only has `RandomStreams(get_rng().randint(1, 2147462579))`: https://github.com/Lasagne/Lasagne/blob/1d1615f/lasagne/layers/noise.py#L74
I thought this would be replaced with a GPU version both in the old and new backend: https://github.com/Lasagne/Lasagne/issues/698#issuecomment-290538224
Is `MRG_RandomStreams` not compatible with the new backend? It's not mentioned as an incompatibility on https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29.

@nouiz or @lamblin, is there anything to change for using random variables in the new gpu backend?

**Comment by jc-Rosichini** [comment]
Hi Jan,
You're right there is none, but 'once in the past' adding use_cuda = True would insure GPU is used.

The same having in .theanorc
device=cuda0
force_device=True

Apparently the switch between CPU mode and GPU mode for RandomStreams has been somewhat altered in the new theano/libgpuarray backend version.
jc

**Comment by f0k** [comment]
> Apparently the switch between CPU mode and GPU mode for RandomStreams has been somewhat altered in the new theano/libgpuarray backend version.

Okay, I can confirm this. I don't think this is intended. Theano/Theano@3767922:
```
>>> import theano
Couldn't import dot_parser, loading of dot files will not be possible.
Using cuDNN version 5103 on context None
Mapped name None to device cuda0: GeForce GTX 970 (0000:01:00.0)
>>> from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
>>> r = RandomStreams()
>>> u = r.uniform((10,20))
>>> theano.printing.debugprint(theano.function([], u))
HostFromGpu(gpuarray) [id A] ''   1
 |GPUA_mrg_uniform{GpuArrayType<None>(float32, (False, False)),inplace}.1 [id B] ''   0
   |<GpuArrayType<None>(int32, (False, False))> [id C]
   |TensorConstant{[10 20]} [id D]
GPUA_mrg_uniform{GpuArrayType<None>(float32, (False, False)),inplace}.0 [id B] ''   0
```
Theano/Theano@388805f:
```
>>> import theano
Couldn't import dot_parser, loading of dot files will not be possible.
Using cuDNN version 5103 on context None
Mapped name None to device cuda0: GeForce GTX 970 (0000:01:00.0)
>>> from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
>>> r = RandomStreams()
>>> u = r.uniform((10,20))
>>> theano.printing.debugprint(theano.function([], u))
mrg_uniform{TensorType(float32, matrix),inplace}.1 [id A] ''   1
 |HostFromGpu(gpuarray) [id B] ''   0
 | |<GpuArrayType<None>(int32, matrix)> [id C]
 |TensorConstant{[10 20]} [id D]
GpuFromHost<None> [id E] ''   2
 |mrg_uniform{TensorType(float32, matrix),inplace}.0 [id A] ''   1
```

**Comment by f0k** [comment]
Bisected this, Theano/Theano@3c4ac6d is the first bad commit. I'll report this upstream.

**Comment by jc-Rosichini** [comment]
nouiz just provided the reason of this.
I just checked his proposed workaround, it's ok.
see: https://github.com/Theano/Theano/issues/5849
cheers

**Comment by f0k** [comment]
Fixed by Theano/Theano#5859. Closing.

Next time please point me to the Theano issue if you open one in parallel! Frédéric already found the solution an hour ago. Would have saved me from bisecting this. But thanks for reporting this anyway!

**Comment by jc-Rosichini** [comment]
Sorry, got feedback from Fred after I warned you about the issue and I was investigating a possible quick an dirty fix.
All things happened in parallel
rgs
jc
